In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Create new Dataframe
df_ords= pd.read_csv(r'C:\Users\Asus\Instacart Basket Analysis\02 Data\Prepared Data\orders_wrangled.csv',index_col=False)

In [3]:
#Check the dataframe of top 5
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [4]:
#Create new dataframe
path=r'C:\Users\Asus\Instacart Basket Analysis\02 Data\Original Data\4.3_orders_products'
path

'C:\\Users\\Asus\\Instacart Basket Analysis\\02 Data\\Original Data\\4.3_orders_products'

In [5]:
df_prods=pd.read_csv(os.path.join(path,'products.csv'),index_col= False)
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [6]:
#Descriptive statistics on product data
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [7]:
#Check for mixed data types
for col in df_prods.columns.tolist():
    weird = (df_prods[[col]].applymap(type)!= df_prods[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_prods[weird])>0:
        print(col)

product_name


## Missing values


In [8]:
# Finding missing values in df_prods

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [42]:
#Check the shape
df_prods.shape

(49677, 5)

In [20]:
df_prods.dropna(inplace=True)

In [22]:
df_prods.shape

(49677, 5)

## Duplicates

In [27]:
#Finding Duplicate

df_dups= df_prods[df_prods.duplicated()]

In [34]:
df_dups.shape

(5, 5)

In [29]:
# Addressing duplicates
df_prods.shape

(49677, 5)

In [30]:
df_prods_clean_no_dups = df_prods.drop_duplicates()

In [32]:
df_prods_clean_no_dups.shape

(49672, 5)

## Task

In [46]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

In [48]:
# Check for mixed data type

for col in df_ords.columns.tolist():
    weird=(df_ords[[col]].applymap(type)!=df_ords[[col]].iloc[0].apply (type)).any(axis=1)
    if len(df_ords[weird])>0:
        print(col)

### There are no mixed data type in df_ords (i.e. Dataframe Orders)

In [50]:
# Check for missing values
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
eval_set                      0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

## Address the missing values using an appropriate method. 

In [78]:
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull()==True]

In [79]:
df_ords_nan 

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_last_order  
0                         2                  8                    NaN  
11                        2                 11                    NaN  
26                        1                 14                    NaN  
39                        6                 11                    NaN  
45                        3                 12                    NaN  
...                     ...                ...                    ...  
3420930                   4                 12                    NaN  
3420934                   3                 18                    NaN  
3421002                   6                 19                    NaN  
3421019                   1                 15                    NaN  
3421069                   3                 11                    NaN  

[206209 rows x 8 columns]

In [80]:
df_ords_nan['order_number'].value_counts(dropna=False)

order_number
1    206209
Name: count, dtype: int64

#### Since these NaN 's are valid as we can see that order number 1 has 206209 count. which means these are genuine first orders so thats why has them as NaN and not any number. We might have to work around using new variables. 

## Duplicate values

In [90]:
df_ords_dups=df_ords[df_ords.duplicated()]

In [91]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

### No Duplicates where found in df_ords

## Exporting product and order data

In [92]:
path = r'C:\Users\Asus\Instacart Basket Analysis\02 Data\Prepared Data'
path

'C:\\Users\\Asus\\Instacart Basket Analysis\\02 Data\\Prepared Data'

In [93]:
df_ords.to_csv(os.path.join(path, 'orders_checked.csv'))

In [94]:
df_prods.to_csv(os.path.join(path,'products_checked.csv'))